stationsim_model/source/Agent.cpp

```diff
@@ -162,46 +166,59 @@ namespace station_sim {
-            location.x = boundaries[1].x;
+    // Assuming `location` is outside of `boundaries`, move it to the closest
+    // point of the region within boundaries
+    void Agent::clip_vector_values_to_boundaries(Point2D &location, std::vector<Point2D> boundaries) {
+        float min_distance = std::numeric_limits<float>::max();
+        Point2D closest_point = location;
+
+        // Loop over all edges of the boundaries (i.e., all segments) and find
+        // the one closest to the point
+        for (std::vector<Point2D>::size_type index = 0; index < boundaries.size() - 1; ++index) {
+            // Extrema of the edge
+            Point2D s1 = boundaries.at(index);
+            Point2D s2 = boundaries.at(index + 1);
+            // Determine distance of location from the edge and its projection
+            // (_within_ the segment), update closest point if necessary
+            std::pair<float,Point2D> this_distance_projection = location.distance_projection(s1, s2);
+            float this_distance = this_distance_projection.first;
+            Point2D this_projection = this_distance_projection.second;
+            if (this_distance < min_distance) {
+                min_distance = this_distance;
+                closest_point = this_projection;
+            }
         }
 
-        if (location.y < boundaries[0].y) {
-            location.y = boundaries[0].y;
-        }
-        if (location.y > boundaries[1].y) {
-            location.y = boundaries[1].y;
-        }
+        // Move location to the closest point
+        location = closest_point;
     }
 
     Point2D Agent::calculate_agent_direction() {
-        float distance = calculate_distance(desired_location, agent_location);
+        float distance = desired_location.distance(agent_location);
 
         return Point2D((desired_location.x - agent_location.x) / distance,
                        (desired_location.y - agent_location.y) / distance);
     }
 
-    float Agent::calculate_distance(Point2D location_0, Point2D location_1) {
-        float sum = powf(location_0.x - location_1.x, 2);
-        sum += powf(location_0.y - location_1.y, 2);
-
-        return std::sqrt(sum);
-    }
-
-    bool Agent::is_outside_boundaries(const std::array<Point2D, 2> boundaries, const Point2D &location) {
-        return location.x < boundaries[0].x || location.x > boundaries[1].x || location.y < boundaries[0].y ||
-               location.y > boundaries[1].y;
+    bool Agent::is_outside_boundaries(const std::vector<Point2D> boundaries, const Point2D &location) {
+        int len = boundaries.size();
+        int j = len - 1;
+        bool c = false;
+        for (int i = 0; i < len; i++) {
+            if (((boundaries[i].y > location.y) != (boundaries[j].y > location.y)) &&
+                (location.x < boundaries[i].x + (boundaries[j].x - boundaries[i].x) * (location.y - boundaries[i].y) /
+                 (boundaries[j].y - boundaries[i].y)))
+                c = !c;
```

Comment:
I spent a long time looking at this before realising it's an implementation of a ray casting algorithm! Looks good, the only thing I could think to suggest was renaming `c` to be more explicit, but the return statement makes it obvious what it's doing.

Reply:
I should have added the comment that this is an implementation of https://en.wikipedia.org/wiki/Even%E2%80%93odd_rule#Implementation